src/nix-build/nix-build.cc

```diff
@@ -0,0 +1,482 @@
+                if (std::regex_search(interpreter, std::regex("perl"))) {
+                        execArgs = "-a PERL";
+                }
+
+                auto joined = stringstream{};
+                for (const auto & i : savedArgs)
+                    joined << shellEscape(i) << ' ';
+
+                if (std::regex_search(interpreter, std::regex("ruby"))) {
+                    // Hack for Ruby. Ruby also examines the shebang. It tries to
+                    // read the shebang to understand which packages to read from. Since
+                    // this is handled via nix-shell -p, we wrap our ruby script execution
+                    // in ruby -e 'load' which ignores the shebangs.
+
+                    envCommand = (format("exec %1% %2% -e 'load(\"%3%\") -- %4%") % execArgs % interpreter % script % joined.str()).str();
+                } else {
+                    envCommand = (format("exec %1% %2% %3% %4%") % execArgs % interpreter % script % joined.str()).str();
+                }
+            }
+
+            else if (!arg.empty() && arg[0] == '-') {
+                buildArgs.push_back(arg);
+            }
+
+            else if (arg == "-Q" || arg == "--no-build-output") {
+                buildArgs.push_back(arg);
+                instArgs.push_back(arg);
+            }
+
+            else {
+                exprs.push_back(arg);
+            }
+        }
+
+        if (packages && fromArgs) {
+            throw UsageError("‘-p’ and ‘-E’ are mutually exclusive");
+        }
+
+        if (packages) {
+            instArgs.push_back("--expr");
+            auto joined = stringstream{};
+            joined << "with import <nixpkgs> { }; runCommand \"shell\" { buildInputs = [ ";
+            for (const auto & i : exprs)
+                joined << '(' << i << ") ";
+            joined << "]; } \"\"";
+            exprs = Strings{joined.str()};
+        } else if (!fromArgs) {
+            if (exprs.empty() && runEnv && access("shell.nix", F_OK) == 0)
+                exprs.push_back("shell.nix");
+            if (exprs.empty())
+                exprs.push_back("default.nix");
+        }
+
+        if (runEnv)
+            setenv("IN_NIX_SHELL", pure ? "pure" : "impure", 1);
+
+        for (auto & expr : exprs) {
+            // Instantiate.
+            auto drvPaths = std::vector<string>{};
+            if (!std::regex_match(expr, std::regex("^/.*\\.drv$"))) {
+                // If we're in a #! script, interpret filenames relative to the
+                // script.
+                if (inShebang && !packages)
+                    expr = absPath(expr, dirOf(script));
+
+                auto instantiateArgs = Strings{"--add-root", drvLink, "--indirect"};
+                for (const auto & arg : instArgs)
+                    instantiateArgs.push_back(arg);
+                instantiateArgs.push_back(expr);
+                auto instOutput = runProgram(settings.nixBinDir + "/nix-instantiate", false, instantiateArgs);
```

Comment:
Why call nix-instantiate as a child process when you can do the evaluation in the same process? Without having to do a lot of ugly argument passing etc.

Likewise for the build step.